### PR TITLE
PICO split and merge

### PR DESCRIPTION
**Pull request by talbrecht**
This pull request adds some special treatment of ice shelves and ocean inputs in the **PICO** ocean module (e.g. in case of a collapse of the West Antarctic Ice Sheet). When ice shelves merge across basin boundaries that both have access to the ocean, but that do not share a common ocean boundary (no basin neighbors at the coast), then those ice shelves are split into two separate instances with individual ocean inputs. When a grounding line retreats into another basins, which does not has a connection to the ocean in that basin (ice shelf cavity), ocean conditions from this basin are not considered in the area-weighted average of the ocean inputs for this ice shelf instance.

**Checklist**

- [x] update documentation
- [x] update [`CHANGES.rst`](CHANGES.rst)
- [x] add regression tests

### CODE REVIEW

File: src/coupler/ocean/PicoGeometry.cc [hunk 1]
```diff
@@ -468,6 +501,167 @@ void PicoGeometry::compute_ocean_mask(const IceModelVec2CellType &cell_type, Ice
...
+        if (M.n == MASK_ICE_FREE_OCEAN or M.e == MASK_ICE_FREE_OCEAN or M.s == MASK_ICE_FREE_OCEAN or M.w == MASK_ICE_FREE_OCEAN) {
+          if (cfs_in_basins_per_shelf[s*m_n_basins+b] != b) {
+            cfs_in_basins_per_shelf[s*m_n_basins+b] = b;
+          }
+        }
+      }
+    }
+
+    //GlobalSum(m_grid->com, cfs_in_basins_per_shelf.data(), m_n_shelves*m_n_basins);
+    //GlobalSum(m_grid->com, n_shelf_cells_per_basin.data(), m_n_shelves*m_n_basins);
```
Comment (ckhroulev) [inline src/coupler/ocean/PicoGeometry.cc:587]: Please remove commented out code to reduce clutter.

File: src/coupler/ocean/PicoGeometry.cc [hunk 2]
```diff
@@ -468,6 +501,167 @@ void PicoGeometry::compute_ocean_mask(const IceModelVec2CellType &cell_type, Ice
...
+      int b = basin_mask.as_int(i, j);
+      int s = shelf_mask.as_int(i, j);
+      int b0 = most_shelf_cells_in_basin[s];
+      if (b != b0 and b != n_basin_neighbors[2*b0] and b != n_basin_neighbors[2*b0+1] and cfs_in_basins_per_shelf[s*m_n_basins+b] > 0) {
+        n_shelf_cells_to_split[s*m_n_basins+b]++;
+      }
+    }
+  }
+
+  //GlobalSum(m_grid->com, n_shelf_cells_to_split.data(), m_n_shelves*m_n_basins);
```
Comment (ckhroulev) [inline src/coupler/ocean/PicoGeometry.cc:631]: Please remove commented out code to reduce clutter.

File: test/test_pico_split/run_test.sh [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+#do
+#  ncap2 -O -s "where(basins==${bi}) salinity_ocean=${si}" input_test01.nc input_test01.nc
+#  bi=$((bi+1))
+#done
+
+#ncap2 -O -s "climatic_mass_balance=0*salinity_ocean" input_test01.nc input_test01.nc
+#ncatted -O -a units,climatic_mass_balance,o,c,"kg m-2 year-1" \
+#           -a long_name,climatic_mass_balance,o,c,"surface mass balance (accumulation/ablation) rate" \
+#           -a standard_name,climatic_mass_balance,o,c,"land_ice_surface_specific_mass_balance_flux" \
+#           input_test01.nc
```
Comment (ckhroulev) [inline test/test_pico_split/run_test.sh:50]: Here and below: please remove commented out code to reduce clutter.

File: src/coupler/ocean/Pico.cc [hunk 4]
```diff
@@ -414,12 +420,33 @@ void Pico::set_ocean_input_fields(const PicoPhysics &physics, const IceModelVec2
...
 
     for (int s = 0; s < m_n_shelves; s++) {
       n_shelf_cells[s] = GlobalSum(m_grid->com, n_shelf_cells[s]);
       for (int b = 0; b < m_n_basins; b++) {
         n_shelf_cells_per_basin[s][b] = GlobalSum(m_grid->com, n_shelf_cells_per_basin[s][b]);
+
+        cfs_in_basins_per_shelf[s][b] = GlobalSum(m_grid->com, cfs_in_basins_per_shelf[s][b]);
+        //if (cfs_in_basins_per_shelf[s][b]>0) {
+        //  m_log->message(2, "PICO, ocean gateways results with s=%d, r=%d \n",s,cfs_in_basins_per_shelf[s][b]);
+        //}
```
Comment (ckhroulev) [inline src/coupler/ocean/Pico.cc:443]: Please remove commented out code to reduce clutter.

**Review by ckhroulev — CHANGES_REQUESTED** [review]
Looks good, but needs a bit of tidying up. Thanks!

File: src/coupler/ocean/Pico.cc (hunk 4 above)
Comment (talbrecht) [reply src/coupler/ocean/Pico.cc:443]: Oops, I guess I rebased merge_pico instead of pico_merge... Will update this one later...

**Comment by talbrecht** [comment]
OK, pik/pico_merge is now rebased to dev, cleaned up old comments, and GlobalSums reformulated for better performance. The test_pico_split has been updated...

**Comment by ckhroulev** [comment]
The current version looks good to me. @talbrecht, please run a couple of tests to check that it still works the way you intended.

And one more thing: we need to automate the regression test. The shell script should return 0 on success and something else (1, for example) on failure (just add `exit 0` and `exit 1` in the right places).

**Comment by talbrecht** [comment]
A added a lot of `if` and `else`, but it seems to work as a quick regression test... I guess the `echo`s could be removed then?

**Comment by ckhroulev** [comment]
> I guess the echos could be removed then?

They aren't hurting anything, so I'd leave it like this.

Thanks!

**Comment by talbrecht** [comment]
The regression test works fine... Always good to learn some more bash...

**Pull request merged**